Reject create table with compact storage

bhalevy opened the pull request:
As discussed in
https://github.com/scylladb/scylladb/issues/12263#issuecomment-1853576813,
compact storage tables are deprecated.

Yet, there's is nothing in the code that prevents users
from creating such tables.

This patch adds a live-updateable config option:
`enable_create_table_with_compact_storage`, set to
`false` by default, that require users to opt-in
in order to create new tables WITH COMPACT STORAGE.

Refs scylladb/scylladb#12263, scylladb/scylladb#16375

* Since this guardrail is an enhancement, no backport is needed

denesb commented on test/cql-pytest/util.py:
```diff
@@ -295,8 +295,18 @@ def __init__(self, cql, key, value):
+            else:
+                raise ValueError(f"Invalid boolean value '{value}'")
+        else:
+            self._original_value = literal_eval(value)
```
> A unit test was added to test/cql-pytest 

I don't see it.

nyh commented on test/cql-pytest/util.py:
```diff
@@ -295,8 +295,18 @@ def __init__(self, cql, key, value):
+                self._original_value = False
+            else:
+                raise ValueError(f"Invalid boolean value '{value}'")
+        else:
```
Why is all these needed? Wouldn't saving the original string work?

bhalevy replied:
no, the current code calls `boost::lexical_cast<bool>` to save the cooked value, but it is very minimalistic and expects `0` or `1`.

nyh commented:
@bhalevy if you disable compact storage by default, how would not all the tests in test/cql-pytest for compact storage fail? Maybe they do, and you just didn't try to run them?

I think you need to add in test/cql-pytest/run.py **and** in test/pylib/scylla_cluster.py the new option to enable compact storage, to allow it to be tested by these tests.

nyh commented:
> As discussed in [#12263 (comment)](https://github.com/scylladb/scylladb/issues/12263#issuecomment-1853576813), compact storage tables are deprecated.

As I commented in that issue now, the "discussion" in that specific comment is based on wrong information that Cassandra deprecated COMPACT STORAGE (does anyone have any source for that?).

We can decide to deprecate COMPACT STORAGE (I won't shed any tears) but let's base it on correct information, and let's update the issue about this deprecation (#12263) when we make such a decision....

And, if we have a patch to deprecate COMPACT STORAGE let's not just refer to #12263 as you did here (you even referred just to a specific comment in it) - let's "Fixes" it.  So let's figure out what needs to be done to Fixes - is this patch enough? This patch and documentation? Something else? If this patch doesn't Fixes this issue, it should mention what's left until the issue is fixed.

bhalevy commented:
@annastuchlik @tzach please review doc change

scylladb-promoter commented:
## Docs Preview :book:
Docs Preview for this pull request is available [here](http://previews.docs.scylladb.com/scylladb/16403)

**Changed Files:**
- [cql/ddl](http://previews.docs.scylladb.com/scylladb/16403/cql/ddl)

**Note:** This preview will be available for 30 days and will be automatically deleted after that period. You can manually trigger a new build by committing changes.

avikivity commented on utils/config_file_impl.hh:
```diff
@@ -202,13 +203,21 @@ void utils::config_file::named_value<T>::set_value(const YAML::Node& node) {
 }
 
+template<>
+bool utils::config_file::lexical_cast<bool>(sstring value);
```
Nit: doesn't need to be a template, overload resolution will pick a non-template over a template.

bhalevy replied:
@avikivity but the function signature has a different return type and same parameter type.
I don't quite follow what you have in mind.

bhalevy replied:
Apparently it does need to be a template, since `utils::config_file::named_value<T>::set_value` calls it like this: `(*this)(lexical_cast<T>(value), src);`

avikivity commented on cql3/statements/create_table_statement.cc:
```diff
@@ -229,6 +229,9 @@ std::unique_ptr<prepared_statement> create_table_statement::raw_statement::prepa
 
     stmt->_use_compact_storage = _properties.use_compact_storage();
+    if (stmt->_use_compact_storage && !db.get_config().enable_create_table_with_compact_storage()) {
+        throw exceptions::invalid_request_exception("CREATE TABLE WITH COMPACT STORAGE support is disabled");
```
Please point to the config value for enabling it, and also mention that it will eventually be removed.

bhalevy replied:
done in v2

avikivity commented on db/config.cc:
```diff
@@ -1119,6 +1119,7 @@ db::config::config(std::shared_ptr<db::extensions> exts)
     , maximum_replication_factor_fail_threshold(this, "maximum_replication_factor_fail_threshold", liveness::LiveUpdate, value_status::Used, -1, "")
     , replication_strategy_warn_list(this, "replication_strategy_warn_list", liveness::LiveUpdate, value_status::Used, {locator::replication_strategy_type::simple}, "Controls which replication strategies to warn about when creating/altering a keyspace. Doesn't affect the pre-existing keyspaces.")
     , replication_strategy_fail_list(this, "replication_strategy_fail_list", liveness::LiveUpdate, value_status::Used, {}, "Controls which replication strategies are disallowed to be used when creating/altering a keyspace. Doesn't affect the pre-existing keyspaces.")
+    , enable_create_table_with_compact_storage(this, "enable_create_table_with_compact_storage", liveness::LiveUpdate, value_status::Used, false, "Enable the deprecated feature of CREATE TABLE WITH COMPACT STORAGE")
```
Mention that it will eventually be removed.

bhalevy replied:
done in v2

avikivity commented on docs/cql/ddl.rst:
```diff
@@ -544,6 +544,11 @@ options of a table are described in the following sections.
+
 ..
    .. caution:: Since Apache Cassandra 3.0, compact tables have the exact same layout internally than non compact ones (for the
       same schema obviously), and declaring a table compact **only** creates artificial limitations on the table definition
```
What's the garbage below "Unchanged files with check annotations Beta"?

bhalevy replied:
I don't understand what you're referring to.
Is it something generated by github or something in ddl.rst?

bhalevy replied:
It looks like github actions is warning about spelling mistakes in comments

mykaul replied:
It does, and you can ignore it.

mykaul replied:
What's less clear to me is how does it still complain on errors I had hoped were fixed by me over multiple PRs :-/

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5231/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6629/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5231/)
#### Failed Tests (89/23477):
- [test_validate_empty_column_name](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5231/testReport/junit/counter_tests/TestCounters/Tests___dtest___test_validate_empty_column_name) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_validate_empty_column_name)
- [test_compact_counter_cluster](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5231/testReport/junit/counter_tests/TestCounters/Tests___dtest___test_compact_counter_cluster) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compact_counter_cluster)
- [test_noncomposite_static_cf](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5231/testReport/junit/cql_additional_tests/TestCQL/Tests___dtest___test_noncomposite_static_cf) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_noncomposite_static_cf)
- [test_dynamic_cf](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5231/testReport/junit/cql_additional_tests/TestCQL/Tests___dtest___test_dynamic_cf) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_dynamic_cf)
- [test_dense_cf](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5231/testReport/junit/cql_additional_tests/TestCQL/Tests___dtest___test_dense_cf) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_dense_cf)
- [test_limit_ranges](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5231/testReport/junit/cql_additional_tests/TestCQL/Tests___dtest___test_limit_ranges) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_limit_ranges)
- [test_limit_multiget](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5231/testReport/junit/cql_additional_tests/TestCQL/Tests___dtest___test_limit_multiget) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_limit_multiget)
- [test_counters](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5231/testReport/junit/cql_additional_tests/TestCQL/Tests___dtest___test_counters) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_counters)
- [test_exclusive_slice](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5231/testReport/junit/cql_additional_tests/TestCQL/Tests___dtest___test_exclusive_slice) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_exclusive_slice)
- [test_in_clause_wide_rows](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5231/testReport/junit/cql_additional_tests/TestCQL/Tests___dtest___test_in_clause_wide_rows) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_in_clause_wide_rows)
#### Build Details:
- Duration: 3 hr 3 min
- Builder: i-017617662b99c978e (m5d.12xlarge)

bhalevy commented:
In v2 ([266d2d7](https://github.com/scylladb/scylladb/commit/266d2d7ce3cc60ac99b586ffa5d911caa99c0a56)):
- option is enabled in db/config but disabled scylla.yaml
  - to reduce churn
  - and allow dtests to pass
  - will be changed to false also in db/config after dtests and SCT are converted
- Added verbiage about removing the feature entirely in the future

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5258/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6659/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5258/)
#### Failed Tests (87/23478):
- [test_validate_empty_column_name](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5258/testReport/junit/counter_tests/TestCounters/Tests___dtest___test_validate_empty_column_name) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_validate_empty_column_name)
- [test_compact_counter_cluster](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5258/testReport/junit/counter_tests/TestCounters/Tests___dtest___test_compact_counter_cluster) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compact_counter_cluster)
- [test_noncomposite_static_cf](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5258/testReport/junit/cql_additional_tests/TestCQL/Tests___dtest___test_noncomposite_static_cf) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_noncomposite_static_cf)
- [test_dynamic_cf](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5258/testReport/junit/cql_additional_tests/TestCQL/Tests___dtest___test_dynamic_cf) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_dynamic_cf)
- [test_dense_cf](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5258/testReport/junit/cql_additional_tests/TestCQL/Tests___dtest___test_dense_cf) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_dense_cf)
- [test_limit_ranges](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5258/testReport/junit/cql_additional_tests/TestCQL/Tests___dtest___test_limit_ranges) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_limit_ranges)
- [test_limit_multiget](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5258/testReport/junit/cql_additional_tests/TestCQL/Tests___dtest___test_limit_multiget) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_limit_multiget)
- [test_counters](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5258/testReport/junit/cql_additional_tests/TestCQL/Tests___dtest___test_counters) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_counters)
- [test_exclusive_slice](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5258/testReport/junit/cql_additional_tests/TestCQL/Tests___dtest___test_exclusive_slice) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_exclusive_slice)
- [test_in_clause_wide_rows](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5258/testReport/junit/cql_additional_tests/TestCQL/Tests___dtest___test_in_clause_wide_rows) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_in_clause_wide_rows)
#### Build Details:
- Duration: 2 hr 15 min
- Builder: spider7.cloudius-systems.com

nyh commented on conf/scylla.yaml:
```diff
@@ -623,3 +623,6 @@ strict_is_not_null_in_views: true
 #  - SimpleStrategy
 # replication_strategy_fail_list:
+
+# Guardrails to enable the deprecated feature of CREATE TABLE WITH COMPACT STORAGE.
```
"Guardrail" in singular.

bhalevy replied:
Will be fixed in v3

nyh commented on cql3/statements/create_table_statement.cc:
```diff
@@ -229,6 +230,16 @@ std::unique_ptr<prepared_statement> create_table_statement::raw_statement::prepa
+            throw exceptions::invalid_request_exception("Support for the deprecated feature of 'CREATE TABLE WITH COMPACT STORAGE' is disabled and will eventually be removed in a future version.  To enable, set the 'enable_create_table_with_compact_storage' config option to 'true'.");
+        }
+        auto msg = "CREATE TABLE WITH COMPACT STORAGE is deprecated and will eventually be removed in a future version.";
+        if (this_shard_id() == 0) {
```
This if() doesn't look right. Unless I'm misunderstanding something, the statement doesn't run on every shard. So why limit the logging just to shard 0?

bhalevy replied:
As far as I saw, `prepare` does run on every shard, so warning on all of them is spamming the log.

nyh commented on test/cql-pytest/test_guardrail_compact_storage.py:
```diff
@@ -0,0 +1,32 @@
+
+def test_create_table_with_compact_storage_default_config(cql, this_dc):
+    ks_defs = f"WITH REPLICATION = {{ 'class' : 'NetworkTopologyStrategy', '{this_dc}' : 1 }}"
+    with new_test_keyspace(cql, ks_defs) as test_keyspace:
```
You're working too hard to create test_keyspace - you can replace all this `this_dc` and `new_test_keyspace()` business by just the existing "test_keyspace" fixture :-) Much shorter, and faster.
You only need to create a new test_keyspace if you're actually testing CREATE KEYSPACE - which you aren't.

bhalevy replied:
Will be fixed in v3

nyh commented on test/cql-pytest/test_guardrail_compact_storage.py:
```diff
@@ -0,0 +1,32 @@
+        with new_test_table(cql, test_keyspace, schema="p int PRIMARY KEY, v int", extra="WITH COMPACT STORAGE") as test_table:
+            pass
+
+@pytest.mark.parametrize("config_value", [False, True])
```
nitpick: instead of calling this parameter "config_value", a name like "enable_compact_storage" would be much more self-explanatory.

bhalevy replied:
Will be fixed in v3

nyh commented on test/cql-pytest/conftest.py:
```diff
@@ -221,3 +221,14 @@ def temp_workdir():
+@pytest.fixture(scope="function")
+def compact_storage(cql):
+    try:
+        with config_value_context(cql, 'enable_create_table_with_compact_storage', 'true') as ctx:
```
This is a nice trick, but it adds a performance penalty (before and after every test you change the configuration) and complexity, and you actually missed one use in test_empty.py (just a single test function).

But I'm curious why you decided to do this instead of just enabliing the `enable_create_table_with_compact_storage` option in test/cql-pytest/run.py. That's what I would have done - and we already do in other places where we need to enable experimental behavior (e.g., user-defined functions).
I see this is exactly what you did for the topology tests (you added the option in test/pylib/scylla_cluster.py) and could have done the same thing for cql-pytest.

bhalevy replied:
> This is a nice trick, but it adds a performance penalty (before and after every test you change the configuration) and complexity, and you actually missed one use in test_empty.py (just a single test function).

Ah, looks like `test_insert_empty_string_key_compact` is marked as `xfail` so I missed it.

About the performance penalty I doubt if it's measurable.
I do prefer the default value in testing to reflect the default in production (i.e. `enable_create_table_with_compact_storage=false`) to make it harder to add new tests that might rely on it - as we often copy-paste existing tests cases when creating new ones.

> 
> But I'm curious why you decided to do this instead of just enabliing the `enable_create_table_with_compact_storage` option in test/cql-pytest/run.py. That's what I would have done - and we already do in other places where we need to enable experimental behavior (e.g., user-defined functions). I see this is exactly what you did for the topology tests (you added the option in test/pylib/scylla_cluster.py) and could have done the same thing for cql-pytest.

The change in `test/pylib/scylla_cluster.py` is needed for the cql tests, not for topology, as I found no wrapper around `cql/cdc_compact_storage_test` for which I could use a fixture to enable compact storage.

bhalevy replied:
As for the performance penalty, I changed the scope of the fixture from `function` to `session` in complete test files testing compact storage in v3.
And also added the `compact_storage` fixture to test_empty.py `test_insert_empty_string_key_compact` and `test_insert_empty_string_compound_clustering_key_compact`.

nyh replied:
I think this change is not a good direction because it relies of pytest not reordering the run order of tests - which you shouldn't rely on.

nyh commented on docs/cql/ddl.rst:
```````````````diff
@@ -544,6 +544,11 @@ options of a table are described in the following sections.
 ``````````````
 
+Support for compact tables has been deprecated.
+By default, newly created tables cannot be defined with the ``COMPACT STORAGE`` option.
```````````````
Idea for slightly different wording (you "create" a table, not "define" one):
"By default, new tables cannot be created with the ``COMPACT STORAGE`` option, but existing tables continue to work as before.
To allow creating new COMPACT STORAGE tables,  set the ..."

bhalevy replied:
fixed wording as per you suggestion in v3.

nyh reviewed (changes requested):
Looks mostly good, but one if() looks fishy to me, and I have a few other comments/requests.

bhalevy commented:
In v3 ([e65dcf8](https://github.com/scylladb/scylladb/commit/e65dcf88a6c7ded92c3f94f90df912b0cdec6311))
- rebased
- config: add enable_create_table_with_compact_storage:
  - Fixed `Guardrail` in singular form in scylla.yaml comment
  - Used `test_keyspace` fixture in test_guardrail_compact_storage.py
- test: enable_create_table_with_compact_storage for tests that need it
  - Used `compact_storage` fixture in test_empty.py
  - Used `scope="session"` in compact_storage_test.py and compact_table_test.py
  - Fixed `test_create_table_with_compact_storage_default_config` in test_guardrail_compact_storage.py
  - Added comment explaining why setting `'enable_create_table_with_compact_storage': True` in test/pylib/scylla_cluster.py is needed.
- docs: ddl: document the deprecation of compact tables
  - Improved phrasing as per @nyh's proposal

bhalevy commented:
unit test failure: https://github.com/scylladb/scylladb/issues/16656#issuecomment-1880056028
dtests need to be adjusted: see https://github.com/scylladb/scylla-dtest/pull/3789

bhalevy commented:
rebased

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6761/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6761/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7988/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6761/)
#### Failed Tests (2/25302):
- [test_alter_inner_udt_by_another_udt](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6761/testReport/junit/user_types_test/TestUserTypes/Tests___dtest___test_alter_inner_udt_by_another_udt) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_alter_inner_udt_by_another_udt)
- [test_add_udt_to_another_data_types](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6761/testReport/junit/user_types_test/TestUserTypes/Tests___dtest___test_add_udt_to_another_data_types) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_add_udt_to_another_data_types)
#### Build Details:
- Duration: 4 hr 57 min
- Builder: i-04f5f147486a874a2 (m5ad.8xlarge)

bhalevy commented:
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6761/) ✅ - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6761/) ❌ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7988/) ✅ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6761/)
> 
> #### Failed Tests (2/25302):
> * [test_alter_inner_udt_by_another_udt](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6761/testReport/junit/user_types_test/TestUserTypes/Tests___dtest___test_alter_inner_udt_by_another_udt) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_alter_inner_udt_by_another_udt)
> * [test_add_udt_to_another_data_types](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6761/testReport/junit/user_types_test/TestUserTypes/Tests___dtest___test_add_udt_to_another_data_types) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_add_udt_to_another_data_types)
> 
> #### Build Details:
> * Duration: 4 hr 57 min
> * Builder: i-04f5f147486a874a2 (m5ad.8xlarge)

https://github.com/scylladb/scylla-dtest/pull/4023

nyh commented:
@bhalevy does this PR Fixes #12263?
Does it, or can it, also Fixes #16375?

bhalevy commented:
> @bhalevy does this PR Fixes #12263?

I don't know since I'm not sure what it requires exactly.

> Does it, or can it, also Fixes #16375?

At this stage is doesn't remove the existing text, just adds caveats that compact storage is disabled by default and how to reenable it, if needed. Plus warn that it will be removed in the future.
I think that we should remove the documentation once we stop supporting the feature entirely.

bhalevy commented:
[f36a3fa](https://github.com/scylladb/scylladb/commit/f36a3fa19e8f6b488455597b2be364e9d9688f14): rebased.
- Ref https://github.com/scylladb/scylladb/issues/20945

> @bhalevy @nyh What blocks merging this PR (after rebasing it)?

I don't think anything blocks it.
@nyh please re-review after last rebase.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:x: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12153/)
#### Build Details:
- Duration: 30 min
- Builder: spider3.cloudius-systems.com

bhalevy commented:
- fixed header dependency regression introduced in previous version

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12167/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12167/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/cql_query_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12167/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12167/)
#### Build Details:
- Duration: 5 hr 7 min
- Builder: spider5.cloudius-systems.com

denesb commented on test/cql-pytest/cassandra_tests/validation/operations/compact_storage_test.py:
```diff
@@ -12,6 +12,12 @@
 compactOption = " WITH COMPACT STORAGE"
 
+# Enable compact_storage for all tests
+@pytest.fixture(scope="session")
```
You are effectively enabling compact storage for the entire test session (for the entire invocation of `pytest` that is.
If you only want it enabled for this one file, use the `module` scope.

See https://docs.pytest.org/en/6.2.x/fixture.html#fixture-scopes.

denesb reviewed (approved):
LGTM (one small comment).

I'd like another review from @nyh as he had comments for the previous version.

bhalevy commented:
In v6 ([3a0463c](https://github.com/scylladb/scylladb/commit/3a0463cc9ace4b124cf649307870f8f09b8845a5)):
- rebased
- changed fixtures scope to `module`

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12199/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12199/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/cql_query_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12199/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4050/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12068/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/993/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12199/)
#### Build Details:
- Duration: 5 hr 11 min
- Builder: spider4.cloudius-systems.com

bhalevy commented:
ping @nyh please re-review and consider merging.

bhalevy commented:
> ping @nyh please re-review and consider merging.

@nyh re-ping.  We'd like to get your blessing for merging this PR.
I've updated https://github.com/scylladb/scylladb/issues/12263#issue-1489820792 with the remaining work and turned it into an epic to be followed up by @paszkow and team.

bhalevy commented:
@nyh ping, please review :point_up:

nyh commented on test/cql-pytest/cassandra_tests/validation/operations/compact_storage_test.py:
```diff
@@ -12,6 +12,12 @@
+# Enable compact_storage for all tests
+@pytest.fixture(scope="module")
+def all_tests_compact_storage(compact_storage):
+    pass
```
I am not happy at all about this approach. It took me a while to understand why this actually works: The "compact_storage" fixture turns on compact storage for the ENTIRE SERVER, and never disables it, so if you run a test with this fixture *first*, all other tests will work. But if pytest reorders the tests, or we do (I think @xtrey is working on doing just that) something like dtest where we deliberately run just a subset of the test functions, this will not work as you hoped it will.

I think you should either use a function-scoped fixture (as in the previous version), or just change the configuration in run.py and test/pytest/scylla_cluster.py. We already do the latter for other deprecated features we want to keep testing and experimental features we want to test - we need to enable them otherwise we can't test them.

xtrey replied:
Test.py triggers the pytest under the hood by file. Pytest by itself does not guarantee the order of the executed tests. So even if it works right now, it's fragile. I totally share the @nyh concerns about this.
However, there is pytest [hook](https://docs.pytest.org/en/stable/reference/reference.html#pytest.hookspec.pytest_collection_modifyitems) where we can modify the order of the tests and put one of the tests as a first one. So we can ensure it will always execute in a first row.

bhalevy replied:
I don't mind solving this issue in a better way that won't require using hooks

bhalevy replied:
But I'd still like to enable compact storage only for functions or complete modules that require it

nyh commented on test/pylib/scylla_cluster.py:
```diff
@@ -137,6 +137,8 @@ def make_scylla_conf(mode: str, workdir: pathlib.Path, host_addr: str, seed_addr
         },
 
+        # Required for cql/cdc_compact_storage_test
+        'enable_create_table_with_compact_storage': True,
```
Now I'm completely confused... If you do this, why do you need all those compact_storage fixture? How can the test test_create_table_with_compact_storage_default_config() pass with this?

By the way, this change in pylib/scylla_cluster.py will only apply to running the tests with test.py. To also affect test/cql-pytest/run, you also need the same option to test/cql-pytest/run.py.

bhalevy replied:
I think I was confused too when doing this.
I don't want this option to be enabled for all tests.
Only for the tests that require it.

bhalevy commented:
In v7 ([1835c21](https://github.com/scylladb/scylladb/commit/1835c216444a2de5b2b48364aa43d4294ca5d116)):
- rebased
- use function-scope compact_storage fixture
- enable compact_storage in `test/cql/suite.yaml` rather than `test/pylib/scylla_cluster.py`

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12944/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12944/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/cql_query_test
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/72/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12613/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4588/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1515/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12944/)
#### Build Details:
- Duration: 5 hr 24 min
- Builder: spider1.cloudius-systems.com

bhalevy commented:
[f8185c3](https://github.com/scylladb/scylladb/commit/f8185c32110e1d51c64bccfef58db617a639ea5c): rebased

@nyh please review before this bit-rots again

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13241/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13241/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/cql_query_test
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/288/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4868/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1779/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12882/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13241/)
#### Build Details:
- Duration: 10 hr
- Builder: i-08c225d8e21eb6e5c (m5ad.8xlarge)

bhalevy commented:
v8 ([bd14b49](https://github.com/scylladb/scylladb/commit/bd14b4965c012ed51e9a630ea09ae1ea5ba99687)): rebased

@nyh ping~2 - please review before this bit-rots yet again

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14371/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14371/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/cql_query_test
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/1088/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5835/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2718/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13846/)
:white_check_mark: - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/490/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14371/)
#### Failed Tests (3/38187):
- [test_view_build_status_snapshot.dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14371/testReport/junit/topology_custom/test_view_build_status/Tests___Unit_Tests___test_view_build_status_snapshot_dev_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_view_build_status_snapshot.dev.3)
- [test_view_build_status_snapshot.dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14371/testReport/junit/topology_custom/test_view_build_status/Tests___Unit_Tests___test_view_build_status_snapshot_dev_3_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_view_build_status_snapshot.dev.3)
- [test_loading_cache_update_config](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14371/testReport/junit/boost/loading_cache_test/test_loading_cache_update_config) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_loading_cache_update_config)
#### Build Details:
- Duration: 10 hr
- Builder: i-0c769c10b86e9bd89 (m5ad.8xlarge)

bhalevy commented:
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14371/) ✅ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14371/) **The following new/updated tests ran 100 times for each mode:** 🔹 boost/cql_query_test ✅ - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/1088/) ✅ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5835/) ✅ - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2718/) ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13846/) ✅ - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/490/) ❌ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14371/)
> 
> #### Failed Tests (3/38187):
> * [test_view_build_status_snapshot.dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14371/testReport/junit/topology_custom/test_view_build_status/Tests___Unit_Tests___test_view_build_status_snapshot_dev_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_view_build_status_snapshot.dev.3)
> * [test_view_build_status_snapshot.dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14371/testReport/junit/topology_custom/test_view_build_status/Tests___Unit_Tests___test_view_build_status_snapshot_dev_3_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_view_build_status_snapshot.dev.3)

Unrelated: https://github.com/scylladb/scylladb/issues/20638

> * [test_loading_cache_update_config](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14371/testReport/junit/boost/loading_cache_test/test_loading_cache_update_config) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_loading_cache_update_config)

Unrelated: https://github.com/scylladb/scylladb/issues/20322

avikivity commented on cql3/statements/alter_table_statement.cc:
```diff
@@ -448,6 +448,8 @@ alter_table_statement::raw_statement::prepare(data_dictionary::database db, cql_
     std::optional<sstring> warning = check_restricted_table_properties(db, s, keyspace(), column_family(), *_properties);
     if (warning) {
+        // FIXME: should this warning be returned to the caller?
+        // See https://github.com/scylladb/scylladb/issues/20945
```
Yes it should be returned

bhalevy replied:
https://github.com/scylladb/scylladb/issues/20945

avikivity commented on cql3/statements/create_table_statement.cc:
```diff
@@ -185,8 +185,17 @@ std::unique_ptr<prepared_statement> create_table_statement::raw_statement::prepa
+    auto stmt_warning = [&] (sstring msg) {
+        if (this_shard_id() == 0) {
+            mylogger.warn("{}: {}", cf_name, msg);
+        }
```
Why is this in prepare and not in execute?

bhalevy replied:
execute is implemented by the super class: schema_altering_statement::execute.
So ai followed the existing design of where the checks are implemented.

nyh replied:
By the way, if we can find an error during prepare (because it's not related to prepared-statement parameters, etc.), isn't it better? Why allow the prepare but then fail the execute? Clearly, for "syntax error", the prepare itself fails, so it's probably good to fail prepare on some of the other types of errors as well.

avikivity commented on test/cqlpy/test_guardrail_compact_storage.py:
```diff
@@ -0,0 +1,28 @@
+import pytest
```
License/copyright blurb

bhalevy replied:
Added in v9

bhalevy commented:
In v9 ([6f633f9](https://github.com/scylladb/scylladb/commit/6f633f9ef347b286a39a22841cdcee3f5696e4b6)):
- rebased (resolved conflicts in db/config.{hh,cc}
- added copyright and license blurb, and file-level comment to new test file

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14465/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14465/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/cql_query_test
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2795/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5915/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13926/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14465/)
#### Build Details:
- Duration: 6 hr 33 min
- Builder: spider6.cloudius-systems.com

bhalevy commented:
[5f011df](https://github.com/scylladb/scylladb/commit/5f011df3e76dfeb756d91f34b024aba445f4806f): rebased

@nyh, this PR keeps rotting while awaiting your final review.
Please review ASAP and let's get it merged.

nyh commented on test/cqlpy/cassandra_tests/validation/operations/compact_storage_test.py:
```diff
@@ -17,7 +17,7 @@
 # See issue #3882
 @pytest.mark.skip
-def testSparseCompactTableIndex(cql, test_keyspace):
+def testSparseCompactTableIndex(compact_storage, cql, test_keyspace):
```
Since **all** the tests in this file are about compact_storage, you could have used an autouse fixture to set this option for all of them instead of each one individually. Perhaps something like this:

```
@pytest.fixture(scope="module", autouse=True)
def all_tests_in_this_file_use_compact_storage(compact_storage):
     pass
```

bhalevy replied:
Done in v10

nyh reviewed (approved):
Looks good. I just have a suggestion how you can use an auto-use fixture instead of adding the "compact_storage" fixture in 20 tests in the same file, but this is just a style improvement. Other than that looks good to me.

I'm not happy about the proliferation of config options that turn features on and off (we seem to be adding those weekly, especially in the context of tablets), but I'll survive :-)

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14564/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14564/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/cql_query_test
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2880/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5999/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14008/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14564/)
#### Failed Tests (18/38621):
- [cqlpy.test_guardrail_compact_storage](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14564/testReport/junit/%28root%29/%28empty%29/Tests___Unit_Tests___cqlpy_test_guardrail_compact_storage) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cqlpy.test_guardrail_compact_storage)
- [cqlpy.test_guardrail_compact_storage](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14564/testReport/junit/%28root%29/%28empty%29/Tests___Unit_Tests___cqlpy_test_guardrail_compact_storage_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cqlpy.test_guardrail_compact_storage)
- [cqlpy.test_guardrail_compact_storage](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14564/testReport/junit/%28root%29/%28empty%29/Tests___Unit_Tests___cqlpy_test_guardrail_compact_storage_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cqlpy.test_guardrail_compact_storage)
- [cqlpy.test_guardrail_compact_storage](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14564/testReport/junit/%28root%29/%28empty%29/Tests___Unit_Tests___cqlpy_test_guardrail_compact_storage_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cqlpy.test_guardrail_compact_storage)
- [cqlpy.test_guardrail_compact_storage](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14564/testReport/junit/%28root%29/%28empty%29/Tests___Unit_Tests___cqlpy_test_guardrail_compact_storage_5) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cqlpy.test_guardrail_compact_storage)
- [cqlpy.test_guardrail_compact_storage](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14564/testReport/junit/%28root%29/%28empty%29/Tests___Unit_Tests___cqlpy_test_guardrail_compact_storage_6) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cqlpy.test_guardrail_compact_storage)
- [cqlpy.test_guardrail_compact_storage](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14564/testReport/junit/%28root%29/%28empty%29/Tests___Unit_Tests___cqlpy_test_guardrail_compact_storage_7) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cqlpy.test_guardrail_compact_storage)
- [cqlpy.test_guardrail_compact_storage](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14564/testReport/junit/%28root%29/%28empty%29/Tests___Unit_Tests___cqlpy_test_guardrail_compact_storage_8) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cqlpy.test_guardrail_compact_storage)
- [cqlpy.test_guardrail_compact_storage](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14564/testReport/junit/%28root%29/%28empty%29/Tests___Unit_Tests___cqlpy_test_guardrail_compact_storage_9) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cqlpy.test_guardrail_compact_storage)
- [cqlpy.test_guardrail_compact_storage.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14564/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cqlpy_test_guardrail_compact_storage_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cqlpy.test_guardrail_compact_storage.debug.1)
#### Build Details:
- Duration: 13 hr
- Builder: i-0d5878fde1fbb7fd4 (m5ad.8xlarge)

bhalevy commented:
In v10 ([5c77956](https://github.com/scylladb/scylladb/commit/5c77956205ce4568b04def60fc7e9749a891b2da)):
- Rebased
- Fixed import in test_guardrail_compact_storage
- Used `autouse=True` for `compact_storage` fixture applied to all tests in the same file.

nyh commented on test/cqlpy/test_guardrail_compact_storage.py:
```diff
@@ -0,0 +1,30 @@
+# Scylla-only. Let's mark them all scylla_only with an autouse fixture:
+@pytest.fixture(scope="function", autouse=True)
+def all_tests_are_scylla_only(scylla_only):
+    pass
```
Nitpick (feel free to ignore, it's not important):
If you have a file with just two tests, it's simpler just to add scylla_only on those two places and not use this autouse and 5-line comment explaining it :-(

A different question for a different time (not important now) is why you even have a test file with just one test ;-)

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14578/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14578/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/cql_query_test
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6011/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2892/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14020/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14578/)
#### Build Details:
- Duration: 13 hr
- Builder: i-00f8e145fcbfa8f33 (m5ad.8xlarge)
